File: 03-Stacks/java/src/MinStack.java
import java.util.Objects;
import java.util.Stack;

public class MinStack {
    private final Stack<Integer> stack = new Stack<>();
    private final Stack<Integer> minStack = new Stack<>();

    public void push(int item) {
        stack.push(item);

        if (minStack.isEmpty())
            minStack.push(item);
        else if (item < minStack.peek())
            minStack.push(item);
    }

    public int pop() {
        if (stack.empty())
            throw new IllegalStateException();

        var top = stack.pop();

        if (Objects.equals(minStack.peek(), top))
            minStack.pop();

        return (int) top;
    }

    public int min() {


        return minStack.peek();
    }


}

File: 03-Stacks/cpp/MinStack.cpp
#include "MinStack.h"

void MinStack::push(int item) {
    st.push(item);
    if (minStack.empty())
        minStack.push(item);
    else if (minStack.top() > item)
        minStack.push(item);

}

int MinStack::pop() {
    if (st.empty())
        throw "stack is empty!";

    int top = st.top();
    st.pop();

    if (top == minStack.top())
        minStack.pop();

    return top;
}

int MinStack::min() {
    return minStack.top();
}


